### ISSUE TITLE
Adding add in iOS device calling build method continuously.

### ISSUE THREAD

**Issue by kmvignesh** [YoussefKababe/admob_flutter#42]
Hi, Today I tried to use this library to have ad on my app. I was working fine on android. But iOS devices was getting hanged all the time. So I did some debugging that time I found that adding "AdmobBanner" causing build method to call continuously.

```
import 'package:admob_flutter/admob_flutter.dart';
import 'package:flutter/material.dart';
import 'dart:io' show Platform;

const AD_MOB_APP_ID_ANDROID = "ca-app-pub-##########~2218177140";
const AD_MOB_BANNER_AD_ID_ANDROID = "ca-app-pub-##########/4022213402";
const AD_MOB_APP_ID_IOS = "ca-app-pub-##########~9065289542";
const AD_MOB_BANNER_AD_ID_IOS = "ca-app-pub-##########/4674440787";

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    Admob.initialize(
        Platform.isAndroid ? AD_MOB_APP_ID_ANDROID : AD_MOB_APP_ID_IOS);
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        accentColor: Colors.blue,
        brightness: Brightness.light,
        appBarTheme: AppBarTheme(color: Colors.blue),
      ),
      home: HomePage(),
    );
  }
}

class AdPage extends StatefulWidget {
  @override
  _AdPageState createState() => _AdPageState();
}

class _AdPageState extends State<AdPage> {
  @override
  Widget build(BuildContext context) {
    print("AdPage build ");
    return Center(
      child: AdmobBanner(
        adUnitId: Platform.isAndroid
            ? AD_MOB_BANNER_AD_ID_ANDROID
            : AD_MOB_BANNER_AD_ID_IOS,
        adSize: AdmobBannerSize.BANNER,
      ),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  @override
  initState() {
    super.initState();
    Admob.initialize(
        Platform.isAndroid ? AD_MOB_APP_ID_ANDROID : AD_MOB_APP_ID_IOS);
  }

  @override
  Widget build(BuildContext context) {
    print("HomePage build ");
    return Scaffold(
      body: Center(
        child: Container(
            child: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Container(
              width: 200.0,
              height: 60.0,
              child: RaisedButton(
                child: Text(
                  "Show Ad",
                  style: TextStyle(fontSize: 21.0, color: Colors.white),
                ),
                color: Theme.of(context).accentColor,
                onPressed: () {
                  Navigator.push(
                      context, MaterialPageRoute(builder: (_) => AdPage()));
                },
                shape: new RoundedRectangleBorder(
                    borderRadius: new BorderRadius.circular(30.0)),
              ),
            ),
          ],
        )),
      ),
    );
  }
}
```

This is the sample project code I created to debug the issue.

Steps:

1. Run the app on iOS device
2. Tap on "Show Ad"
3. Now the device will get hang. you can see the log, where it will print "AdPage build" too many times.

**Comment by kmcgill88** [comment]
Thanks for the detailed bug report @kmvignesh! I have an app using this on iOS and it too will occasionally "hang", but it is inconsistent. It might be a second or two but then continues to render. Are you experiencing that too or does the app just lock up forever?

**Comment by kmvignesh** [comment]
Sorry for the delay. It is happening always for me. I tried to include that on my app. It was hanging after 2-4 sec. I used the same code on new project. In that project also its hanging.

**Comment by diruuu** [comment]
I experience this too. In my case it's because I'm using scope_models. If I remove the scope_models listeners, the continuously rebuilding issue won't happen. I don't why. This is the only reason why I stop using this package. But flutter doesn't have any alternative to show admob as widget, so I guess I need to wait for solution here.

**Comment by CosmoInSpace** [comment]
@diruuu i can back this up with the `provider` package, same story.

**Comment by CosmoInSpace** [comment]
> I experience this too. In my case it's because I'm using scope_models. If I remove the scope_models listeners, the continuously rebuilding issue won't happen. I don't why. This is the only reason why I stop using this package. But flutter doesn't have any alternative to show admob as widget, so I guess I need to wait for solution here.

i can back this up with the provider package, same story.

**Comment by diruuu** [comment]
I think I found a temporary solution for this. I'm wrapping my widget with `fragment` package. What it does basically to prevent rerendering just like `shouldComponentUpdate` does in ReactJS. You should check it out https://pub.dev/packages/fragment

**Comment by CosmoInSpace** [comment]
> I think I found a temporary solution for this. I'm wrapping my widget with `fragment` package. What it does basically to prevent rerendering just like `shouldComponentUpdate` does in ReactJS. You should check it out https://pub.dev/packages/fragment

@diruuu this actually might work, would you mind providing a gist?

**Comment by diruuu** [comment]
> > I think I found a temporary solution for this. I'm wrapping my widget with `fragment` package. What it does basically to prevent rerendering just like `shouldComponentUpdate` does in ReactJS. You should check it out https://pub.dev/packages/fragment
> 
> @diruuu this actually might work, would you mind providing a gist?
> possibly related to [#18 (comment)](https://github.com/YoussefKababe/admob_flutter/issues/18#issuecomment-506290310)

I just simply wrap my widget with `fragment` widget.
```
return fragment((prev, prevKeys) { // previous result & previous keys. Both null on the first run
   return Text(text); // widgets subtree to cache
}, keys: [text]);
```
It does prevent the reloading, but you have to put in on the parent widget, and because of that now I need to compare all keys, which I think will be error prone since I'm using scope_model and I don't want my app stop being reactive. If only there's a way to stop admob_flutter only from being reloaded.

Do you find other solution for this? @CosmoInSpace

**Comment by CosmoInSpace** [comment]
> > > I think I found a temporary solution for this. I'm wrapping my widget with `fragment` package. What it does basically to prevent rerendering just like `shouldComponentUpdate` does in ReactJS. You should check it out https://pub.dev/packages/fragment
> > 
> > 
> > @diruuu this actually might work, would you mind providing a gist?
> > possibly related to [#18 (comment)](https://github.com/YoussefKababe/admob_flutter/issues/18#issuecomment-506290310)
> 
> I just simply wrap my widget with `fragment` widget.
> 
> ```
> return fragment((prev, prevKeys) { // previous result & previous keys. Both null on the first run
>    return Text(text); // widgets subtree to cache
> }, keys: [text]);
> ```
> 
> It does prevent the reloading, but you have to put in on the parent widget, and because of that now I need to compare all keys, which I think will be error prone since I'm using scope_model and I don't want my app stop being reactive. If only there's a way to stop admob_flutter only from being reloaded.
> 
> Do you find other solution for this? @CosmoInSpace

No, this using `fragments` in my case, doesn't work. Not sure if I'm doing something wrong (not using keys for instance) although after some digging I think there needs to be an added pull request with `AutomaticKeepAliveClientMixin` as mentioned in https://github.com/YoussefKababe/admob_flutter/issues/18#issuecomment-506290310 and https://github.com/dreamsoftin/facebook_audience_network/commit/d29d594743cdc2c454c9db81c76e63ab8f8d124c.
Any opinions? @diruuu @kmcgill88

**Comment by elowareKf** [comment]
I solved this for me using a StatefulWidget as a container:

```
class TestContainer extends StatefulWidget{
  @override
  State<StatefulWidget> createState() => TestContainerState();

}

class TestContainerState extends State<TestContainer>{
  Widget banner;

  @override
  void initState() {

    super.initState();

    banner = AdHelper.getBanner();
  }

  @override
  Widget build(BuildContext context) {
    return banner;
  }

}
```

and then using the TestContainer.
